File: sdk/java/src/test/java/io/ketebe/RealServerTest.java
package io.ketebe;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.math.BigInteger;
import java.util.List;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.condition.EnabledIfEnvironmentVariable;
import static org.junit.jupiter.api.Assertions.*;

class RealServerTest {
    @Test
    @EnabledIfEnvironmentVariable(named = "KETEBE_BASE_URL", matches = ".+")
    void typed_ids_and_query_round_trip_against_real_server() {
        String base = System.getenv("KETEBE_BASE_URL");
        KetebeClient client = new KetebeClient(base);
        String collection = "java-sdk-" + System.nanoTime();
        ObjectNode create = new ObjectMapper().createObjectNode();
        create.put("id", collection); create.put("dimension", 2); create.put("metric", "dot");
        client.createCollection(create);
        try {
            client.batchUpsertRecords(collection, List.of(
                new BatchRecordUpsert(RecordId.string("42"), new RecordUpsert(List.of(1.0, 0.0), null)),
                new BatchRecordUpsert(RecordId.unsigned(BigInteger.valueOf(42)), new RecordUpsert(List.of(1.0, 0.0), null))
            ));
            QueryResponse response = client.query(collection, new QueryRequest().vector(List.of(1.0, 0.0)).topK(2).execution("exact"));
            assertEquals(2, response.hits().size());
            assertTrue(response.hits().stream().anyMatch(hit -> hit.id().equals(RecordId.string("42"))));
            assertTrue(response.hits().stream().anyMatch(hit -> hit.id().equals(RecordId.unsigned(BigInteger.valueOf(42)))));
            assertTrue(response.hits().stream().allMatch(hit -> hit.sequenceNumber().signum() >= 0));
        } finally { client.deleteCollection(collection); }
    }
}

File: sdk/go/real_server_test.go
package ketebe

import (
	"context"
	"os"
	"testing"
)

func TestRealServerTypedIDsAndQuery(t *testing.T) {
	base := os.Getenv("KETEBE_BASE_URL")
	if base == "" {
		t.Skip("KETEBE_BASE_URL not set")
	}
	c, err := NewClient(ClientOptions{BaseURL: base})
	if err != nil {
		t.Fatal(err)
	}
	ctx := context.Background()
	_, err = c.CreateCollection(ctx, CreateCollection{ID: "go-sdk", Dimension: 2, Metric: "dot"})
	if err != nil {
		t.Fatal(err)
	}
	defer c.DeleteCollection(ctx, "go-sdk")

	_, err = c.BatchUpsertRecords(ctx, "go-sdk", BatchUpsert{Records: []BatchRecordUpsert{
		{ID: StringID("42"), Vector: []float32{1, 0}},
		{ID: U64ID(42), Vector: []float32{1, 0}},
	}})
	if err != nil {
		t.Fatal(err)
	}

	out, err := c.Query(ctx, "go-sdk", QueryRequest{Vector: []float32{1, 0}, TopK: 2, Execution: "exact"})
	if err != nil {
		t.Fatal(err)
	}
	if len(out.Hits) != 2 {
		t.Fatalf("hits=%d", len(out.Hits))
	}
	seenString, seenU64 := false, false
	for _, hit := range out.Hits {
		if hit.ID == StringID("42") {
			seenString = true
		}
		if hit.ID == U64ID(42) {
			seenU64 = true
		}
	}
	if !seenString || !seenU64 {
		t.Fatalf("typed ids not preserved: %#v", out.Hits)
	}
}
